src/alexandria3k/data_sources/uspto.py

```diff
@@ -18,71 +18,581 @@
+            with zipfile.ZipFile(uspto_record, "r") as zip_ref:
+                # Pick only the XML file inside zip
+                xml_file = [
+                    file
+                    for file in zip_ref.namelist()
+                    if file.endswith(".xml")
+                ]
+                (self.file_name,) = xml_file
+
+                # Get chunks of concatenated XML files
+                with zip_ref.open(self.file_name, "r") as uspto_weekly_file:
+                    # perf.log("Decoding and splitting weekly XML into chunks...")
+
+                    xml_content = uspto_weekly_file.read().decode("utf-8")
+                    xml_files = xml_content.split(XML_DELIMITER)
+
+                    for self.xml_info in xml_files:
+                        if len(self.xml_info) == 0:
+                            continue
+
+                        # Get container name through regex
+                        match = re.search(pattern, self.xml_info[:200])
+                        extracted_id = match.group(1)
+
+                        self.patent_file_name.append(extracted_id)
+                        self.unique_patent_xml_files.append(self.xml_info)
+
+    def get_xml_files_array(self):
+        """Return the array of data files"""
+        return self.unique_patent_xml_files
+
+    def get_container_iterator(self):
+        """Return an iterator over the int identifiers of all data files"""
+        return range(0, len(self.unique_patent_xml_files))
+
+    def get_container_name(self, fid):
+        """Return the name of the file corresponding to the specified fid"""
+        return self.patent_file_name[fid]
+
+
+class VTSource:
+    """Virtual table data source.  This gets registered with the apsw
+    Connection through createmodule in order to instantiate the virtual
+    tables."""
+
+    def __init__(self, data_directory, sample):
+        self.data_files = ZipFiles(data_directory, sample)
+        self.table_dict = {t.get_name(): t for t in tables}
+        self.sample = sample
+
+    def Create(self, _db, _module_name, _db_name, table_name):
+        """Create the specified virtual table
+        Return the tuple required by the apsw.Source.Create method:
+        the table's schema and the virtual table class."""
+        table = self.table_dict[table_name]
+        # print(self.data_files)
+        return table.table_schema(), StreamingCachedContainerTable(
+            table,
+            self.table_dict,
+            self.data_files.get_xml_files_array(),
+            self.sample,
+        )
+
+    Connect = Create
+
+    def get_container_iterator(self):
+        """Return an iterator over the data files identifiers"""
+        return self.data_files.get_container_iterator()
+
+    def get_container_name(self, fid):
+        """Return the name of the file corresponding to the specified fid"""
+        return self.data_files.get_container_name(fid)
+
+
+class XMLChunkCursor:
```

Comment:
Consider moving the many common elements with CrossRef `FilesCursor` to a parent class in `data_source.py` and subclassing.

Reply:
Created an ItemsCursor class on data_sources.py. I made the corresponding changes on both USPTO and Crossref, plus tested them.